calyx/src/analysis/read_write_set.rs

```diff
@@ -18,6 +18,60 @@ impl ReadWriteSet {
             .filter(|port| !port.borrow().is_hole())
     }
 
+    /// Returns the "meaningful" [ir::Port] which are read from in the assignments.
+    /// "Meaningful" means we just exclude the following `@done` reads:
+    /// the `@go` signal for the same cell *must* be written to in the group
+    pub fn meaningful_port_read_set<'a>(
+        assigns: impl Iterator<Item = &'a ir::Assignment> + Clone + 'a,
+    ) -> impl Iterator<Item = RRC<ir::Port>> + 'a {
+        // go_writes = all cells which are guaranteed to have their go port written to in assigns
+        let go_writes: Vec<RRC<ir::Cell>> =
+            Self::port_write_set(assigns.clone().filter(|asgn| {
+                // to be included in go_writes, one of the following must hold:
+                // a) guard is true
+                // b) cell.go = !cell.done ? 1'd1
+                if asgn.guard.is_true() {
+                    return true;
+                }
+                // shares common code with the group2seq pass. Might be a good idea
+                // address this at some point
+                let guard_not_done = |guard: &ir::Guard| -> bool {
```

Comment:
Definitely worth defining this as a function somewhere instead of reimplementing it. Maybe even on the `Guard` type to check if something is a not of a provided `ir::Guard::Port`

Reply:
I added a method on the guard type-- although it takes a cell name of type `ir::Id` instead of an `ir::Guard::Port`. I could make it take an `ir::Guard::Port`, it would just take refactoring the group2seq pass to work with `RRC<Cell>` instead of cell names of type `ir::Id`. If you think this would be a good change to make then lmk.